.github/workflows/main.yml

```diff
@@ -1,50 +1,120 @@
-        uses: actions/setup-python@v2
+          # ref: ${{ github.event.pull_request.head.sha }}
+          fetch-depth: 0  # To get changed lines for coverage-diff: origin/master..HEAD
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
         with:
-          python-version: '2.7'
+          python-version: ${{ matrix.python-version }}
 
       - name: Install dependencies
         run: |
-          python -m pip install --upgrade pip
-          pip install -r requirements-dev.txt
-          # FIXME: branding.py still has no permanent home
-          curl https://gist.github.com/ydirson/3c36a7e19d762cc529a6c82340894ccc/raw/5ca39f621b1feab813e171f535c1aad1bd483f1d/branding.py -O -L
-          pip install pyliblzma
-          pip install -e .
-          command -v xz
-
-      - name: Test
+          python -m pip install --quiet --upgrade pip
+          pip install --quiet -r requirements-dev.txt coveralls # FIXME:
+          if [ ${{ matrix.python-version }} = 2.7 ]; then python2 setup.py install --user;
+          else pip install --quiet -e .;fi
+
+      - name: Post a pylint Report on the Action's summary page
+        if: ${{ matrix.python-version == '3.9' }}
+        run: |
+          pip install --quiet pandas tabulate
+          tool=github-action-pylint-report.py
+          repo=$GITHUB_SERVER_URL/$GITHUB_REPOSITORY
+          tree=$repo/tree/$GITHUB_REF_NAME
+          ./github-action-pylint-report.py xcp pylint-report.md $tree
+          message_count=`tail -1 pylint-report.md|cut -d'|' -f2`
+          if [ -e $GITHUB_STEP_SUMMARY ]; then  #  The "act" local runner does not have this:
+              echo -n "### [$tool]($tree/$tool) on **xcp/\\*\\*/*.py**" >>$GITHUB_STEP_SUMMARY
+              echo ": #Total Pylint messages: $message_count :rocket: " >>$GITHUB_STEP_SUMMARY
+              cat pylint-report.md                                      >>$GITHUB_STEP_SUMMARY
+          fi
+
+      - name: Run pytest with --cov and generate HTML coverage reports
+        id: coverage
+        if: ${{ matrix.python-version == '3.11' }}
         run: |
-          pytest --cov -rP
+          set -x;PYTHONPATH=$PWD pytest --cov
           coverage xml
-          coverage html
-          coverage html -d htmlcov-tests --include="tests/*"
-          diff-cover --html-report coverage-diff.html coverage.xml
+          if [ -n "$COVERALLS_REPO_TOKEN" ]; then coveralls --submit=coverage.xml; fi
+          coverage html -d htmlcov-${{ matrix.python-version }}
+          coverage html -d htmlcov-tests-${{ matrix.python-version }} --include="tests/*"
+          diff-cover --compare-branch=origin/master --html-report coverage-diff-${{ matrix.python-version }}.html coverage.xml
+        env:
+          COVERALLS_REPO_TOKEN: ${{ secrets.COVERALLS_REPO_TOKEN }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: "Run tox for ${{ matrix.python-version }} (and also mypy if 3.11)"
```

Comment:
If tox is used to test on several versions I'd rather drop all the versions from the github actions matrix and just use the latest one

Reply:
poetry is used by tox to setup the py<number> virtual environments.